rfd/0098-registered-openssh-nodes.md

```diff
@@ -0,0 +1,113 @@
+
+### Agentless CA
+
+For security related reasons that will be discussed below, a new CA will be added called Agentless CA. The Agentless CA will be responsible for generating and signing user certificates that will be used to authenticate with registered OpenSSH nodes. The public key of the Agentless CA will be copied to all registered OpenSSH nodes and configured as `TrustedUserCAKeys` in `sshd_config`.
+
+### RBAC
+
+When OpenSSH nodes are registered currently, RBAC checks for those nodes are not performed. RBAC logic will have to be updated so RBAC checks for registered OpenSSH nodes are performed. There are multiple ways to implement this, each with their own advantages and disadvantages.
+
+#### Option 1
+
+Each registered OpenSSH node will have a unique Agentless CA. The CA public key will include information about the node that it was created for, including the node's UUID and the node's labels. When `(lib/srv.AuthHandlers).UserKeyAuth` is called to authenticate a node's user certificate, it will check if node information is present in the CA public key. If node information is present, an RBAC check will be performed using the label information in the CA public key.
+
+Pros:
+
+- No need to lookup node resources to perform RBAC checks
+
+Cons:
+
+- Requires connecting to node to perform RBAC check
+- Updating node labels requires generating and distributing a new host key
+- CA rotation is more complex as every registered OpenSSH node requires a unique CA
+
+#### Option 2
+
+When a user sends a request to a Proxy to connect to a node, the Proxy will attempt to find a node resource by either its hostname or IP, whichever the user specified. If the resource exists and has the `agentless` `sub_kind`, an RBAC check will be performed. If the resource does not exist or isn't an `agentless` node, the connection flow will continue as normal.
+
+Pros:
+
+- No need to connect to node to perform RBAC check
+- Updating node labels can simply be done by using `tctl`
+- CA rotation is simple as every registered OpenSSH node will have the same CA
+
+Cons: none (that I can think of)
+
+#### My choice: Option 2
+
+Option 2 makes CA rotation and updating node labels much easier, and has the only downside of having to search for node resources before connecting to one. 
+
+### Security
+
+Both RBAC checks and session recording for registered OpenSSH nodes require that users connect through a Proxy. If users are able to connect to registered OpenSSH nodes directly, they can bypass both features. Currently the User CA public key is copied to OpenSSH nodes and configured to be trusted by `sshd`. This means OpenSSH nodes will accept certificates that are signed by the User CA. The problem with that is when Teleport users authenticate with a cluster, the Auth server replies with a certificate that is signed by the User CA. Users could potentially use this certificate to directly connect to registered OpenSSH nodes.
```

Comment:
Can you add a section explaining "migration" process i.e. how would users who already have a fleet of OpenSSH servers migrate them to this "agentless" mode.

Reply:
That's a good idea, haven't thought of that. One thing to note is currently Agentless discovery and registration is limited to AWS, so this will only work for users with OpenSSH servers there. I imagine we may want to expand Agentless discovery and registration to work on other cloud platforms, but I guess the underlying cloud platform shouldn't affect how agentless nodes are registered and used.

Reply:
@r0mant Do we want to eventually migrate "proxy recording" users to use agentless mode? I thought that they were different features, and we do expect people to move. 
If we want to give such an option, can we work on that in a different RFD? I think that the scope of that task is huge (updating the current OpenSSH installation, creating RBAC rules, potentially managing AWS and non-AWS nodes that should still work in the "old" mode, etc).

Reply:
@capnspacehook Can you describe the migration process? A user must register all nodes in Teleport and replace the trusted CA.

Reply:
Done